### PR TITLE
[FAB-17786] Update upgrade_dbs peer command to drop state couchdb

### PR DESCRIPTION
**Pull request by wenjianqiao**
Signed-off-by: Wenjian Qiao <wenjianq@gmail.com

#### Type of change
- Improvement (improvement to code, performance, etc)

#### Description
The existing upgrade_dbs command does not automatically drop state couchdbs
and therefore a separate step is required to drop couchdbs, This PR updates
the command to automatically drop state couchdbs.

#### Related issues
https://jira.hyperledger.org/browse/FAB-17786

#### Release Note
TBD

### CODE REVIEW

**Comment by wenjianqiao** [comment]
Hit ci flake FAB-16233

**Comment by wenjianqiao** [comment]
ci-run

File: core/ledger/kvledger/upgrade_dbs.go [hunk 1]
```diff
@@ -24,29 +32,39 @@ func UpgradeDBs(rootFSPath string) error {
...
 	logger.Infof("Ledger data folder from config = [%s]", rootFSPath)
+	dbPath := LedgerProviderPath(rootFSPath)
+	db := leveldbhelper.CreateDB(&leveldbhelper.Conf{DBPath: dbPath})
+	db.Open()
+	defer db.Close()
 
-	// upgrades idStore (ledgerProvider) and drop databases
-	if err := UpgradeIDStoreFormat(rootFSPath); err != nil {
+	idStore := &idStore{db, dbPath}
+	format, err := idStore.getFormat()
```
Comment (cendhu) [inline core/ledger/kvledger/upgrade_dbs.go]: before checking for the format, shouldn't we check whether the `idStore` is empty and return early. It is a remote case though.

File: core/ledger/kvledger/upgrade_dbs_test.go [hunk 2]
```diff
@@ -62,22 +54,104 @@ func TestUpgradeDBs(t *testing.T) {
 	require.Equal(t, os.IsNotExist(err), true)
 
 	// upgrade again should be successful
-	err = UpgradeDBs(rootFSPath)
+	err = UpgradeDBs(conf)
 	require.NoError(t, err)
 }
 
-func TestUpgradeIDStoreWrongFormat(t *testing.T) {
+func TestUpgradeDBsWithV13CouchDB(t *testing.T) {
```
Comment (cendhu) [inline core/ledger/kvledger/upgrade_dbs_test.go]: For rollback and reset commands, we have added tests in the ledger-level integration test pkg. This is because these commands modify blockstore and many DBs (affects multiple sub-packages in the ledger). The upgradeDB command is also modifying blockstore and manyDBs. Further, `TestUpgradeDBsWithV13CouchDB()` and `TestUpgradeDBsWithV13CouchDB()` access testdata present in the ledger-level integration test pkg. Hence, I would prefer moving these tests to `kvledger/tests` to maintain consistency. 

This reminds me of the discussion @manish-sethi and I had a year back on moving recovery-related tests to `kvledger/tests` pkg as it involves multiple sub-packages in the ledger.

File: core/ledger/kvledger/upgrade_dbs.go [hunk 3]
```diff
@@ -24,29 +32,39 @@ func UpgradeDBs(rootFSPath string) error {
...
-	logger.Debugf("Attempting to upgrade idStore data format to current format %s", dataformat.Version20)
+	logger.Info("Updating the ledger provider database format ...")
+	return idStore.upgradeFormat()
+}
 
-	dbPath := LedgerProviderPath(rootFSPath)
-	db := leveldbhelper.CreateDB(&leveldbhelper.Conf{DBPath: dbPath})
-	db.Open()
-	defer db.Close()
+func dropCouchAndLevelDBs(config *ledger.Config) error {
```
[the remark this answers is not in the document]
Comment (wenjianqiao) [reply core/ledger/kvledger/upgrade_dbs.go]: Prefer to use a separate PR for reset/rollback because they have user impact. As of dropDBs unnecessarily drops stateleveldb in the case of couchdb, it is true for other commands too. I can implement a new method instead of calling dropDBs.

File: core/ledger/kvledger/upgrade_dbs_test.go (hunk 2 above)
Comment (wenjianqiao) [reply core/ledger/kvledger/upgrade_dbs_test.go]: Thank you, Senthil. Agree that we should have ledger-level integration tests for UpgradeDBs under kvledger/tests. However, these tests are meant for unit tests to verify idStore is updated correctly and databases are dropped or not dropped as expected.

File: core/ledger/kvledger/upgrade_dbs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (manish-sethi) [reply core/ledger/kvledger/upgrade_dbs.go]: IMO, better to leave `dropDBs` as is. Instead, rename drop `dropStateLevelDB` as `dropStateDB`. Then, within `dropStateDB`, you can check and delete one of the leveldb or couchdb.

File: core/ledger/kvledger/upgrade_dbs.go [hunk 4]
```diff
@@ -24,29 +32,39 @@ func UpgradeDBs(rootFSPath string) error {
...
-	if err := UpgradeIDStoreFormat(rootFSPath); err != nil {
+	idStore := &idStore{db, dbPath}
+	format, err := idStore.getFormat()
+	if err != nil {
 		return err
 	}
+	if bytes.Equal(format, []byte(dataformat.Version20)) {
+		logger.Info("Ledger data format is current, do nothing.")
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (wenjianqiao) [reply core/ledger/kvledger/upgrade_dbs.go]: I recall that we purposely return an error if the idStore format is neither nil (1.x) or 2.0 with a forward thinking. When 3.0 if available, the method will only upgrade from 2.0 to 3.0 and return an error for other formats. Given it, idStore.hasLastestFormat() may not be helpful if there are more than 2 formats. Of course, we can simplify the code for now since we only have 1.x and 2.0.
@manish-sethi what do you think?

File: core/ledger/kvledger/kv_ledger_provider.go [hunk 5]
```diff
@@ -498,6 +498,10 @@ func openIDStore(path string) (s *idStore, e error) {
 	return &idStore{db, path}, nil
 }
 
+func (s *idStore) getFormat() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (manish-sethi) [reply core/ledger/kvledger/kv_ledger_provider.go]: I am not always against the tiny function. In this case, it does not create a problem. I feel this problematic when a logic that makes sense in a single flow is broken down in multiple tiny functions and eventually all those functions are just called one after another and further, in isolation, mostly, none of those functions justify their existence.

File: core/ledger/kvledger/upgrade_dbs_test.go (hunk 2 above)
Comment (cendhu) [reply core/ledger/kvledger/upgrade_dbs_test.go]: In `kvledger/tests`, we do have [verifyRebuilableDoesNotExist()](https://github.com/hyperledger/fabric/blob/master/core/ledger/kvledger/tests/env.go#L144) and [verifyRebuilablesExist()](https://github.com/hyperledger/fabric/blob/master/core/ledger/kvledger/tests/env.go#L126) which checks whether the DBs has been dropped or DBs exist. We use these in both reset and rollback tests. These functions can be extended to add a CouchDB check too. You can find no UT for rollback and reset commands in `kvledger` package. 

As I mentioned in one of the above comments, if we have `hasVersion20Format()`, we can also add an exported function to `test_export.go` and use it in the `kvledger/tests` to check whether the format has been applied.

We can discuss further and can be done in a separate PR too.

File: core/ledger/kvledger/upgrade_dbs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cendhu) [reply core/ledger/kvledger/upgrade_dbs.go]: The suggestion given by Manish seems better.

File: core/ledger/kvledger/upgrade_dbs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cendhu) [reply core/ledger/kvledger/upgrade_dbs.go]: I am not questioning the return of an `error`. I mentioned the side-effect in upgradeFormat and lines 41 to 48 occur twice overall. The `updateFormat()` is in fact doing `updateFormatIfCurrentFormatIsNotV20()`. If we can split that into two, we can avoid this code duplication, execution duplication as well as the side-effect.

File: core/ledger/kvledger/upgrade_dbs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wenjianqiao) [reply core/ledger/kvledger/upgrade_dbs.go]: Yes, agree with Manish's suggestion, but it means that we will drop statecouchdb for all reset/rollback/upgrade commands. Meanwhile, I try to limit the scope of this PR. Maybe we can use this PR to update dropDBs (and dropStateDB) for all 3 commands and use a separate PR to update integration tests. What do you think?

[the remark this answers is not in the document]
Comment (cendhu) [reply core/ledger/kvledger/upgrade_dbs.go]: Yes, we can limit the scope of this PR as it simplifies the reviewer's task too.

File: core/ledger/kvledger/upgrade_dbs.go (hunk 1 above)
Comment (wenjianqiao) [reply core/ledger/kvledger/upgrade_dbs.go]: Good point. This is like a user error. idStore cannot be empty unless something is wrong, e.g., ledger is corrupted or fileSystemPath in core.yaml points to a wrong path.

**Comment by wenjianqiao** [comment]
@cendhu thank you for your comments. I have addressed most of your comments except that I didn't rename idStore getFormat to hasVersion20Format() because I want to avoid adding such a method for each new format.

File: core/ledger/kvledger/upgrade_dbs.go [hunk 6]
```diff
@@ -24,29 +33,44 @@ func UpgradeDBs(rootFSPath string) error {
...
+	dbPath := LedgerProviderPath(rootFSPath)
+	db := leveldbhelper.CreateDB(&leveldbhelper.Conf{DBPath: dbPath})
+	db.Open()
+	defer db.Close()
 
-	// upgrades idStore (ledgerProvider) and drop databases
-	if err := UpgradeIDStoreFormat(rootFSPath); err != nil {
-		return err
+	if emptydb, _ := db.IsEmpty(); emptydb {
+		return errors.Errorf("ledger database %s is empty. Make sure core.yaml has the correct ledger path.", dbPath)
```
Comment (cendhu) [inline core/ledger/kvledger/upgrade_dbs.go]: it will be empty when the peer has joined no channel. Anyway, as I mentioned earlier, it is a remote corner case.

**Review by cendhu — COMMENTED** [review]
Thank you, @wenjianqiao,  for the changes. Overall, LGTM. I yet to look at tests in detail. 

> I didn't rename idStore getFormat to hasVersion20Format()

This is completely fine. My major concern was that the following code was duplicated and executed twice (+ some side-effect). In the latest commit, you have fixed this issue. We can even remove the `getFormat()` as it used at only one place and move `s.db.Get()` to the place where it calls `getFormat()`. Don't need to push anything until Manish reviews this PR. 

```
        format, err := s.db.Get(formatKey)
	if err != nil {
		return err
	}
	idStoreFormatBytes := []byte(dataformat.Version20)
	if bytes.Equal(format, idStoreFormatBytes) {
		logger.Debug("Format is current, nothing to do")
		return nil
	}
	if format != nil {
		err = &dataformat.ErrVersionMismatch{
			ExpectedVersion: "",
			Version:         string(format),
			DBInfo:          fmt.Sprintf("leveldb for channel-IDs at [%s]", s.dbPath),
		}
		return err
	}
```

File: core/ledger/kvledger/tests/v1x_test.go [hunk 7]
```diff
@@ -107,18 +82,45 @@ func TestV13WithStateCouchdb(t *testing.T) {
 	dataHelper.verify(h2)
 
 	// drop couchDB and other rebuildable dbs to test db rebuild
-	dropCouchDBs(t, couchdbConfig)
+	require.NoError(t, statecouchdb.DropApplicationDBs(couchdbConfig))
```
Comment (manish-sethi) [inline core/ledger/kvledger/tests/v1x_test.go]: Not in this PR but just noticed - though we are not writing anything in parallel here, still better to move this line after the next line (i.e., after closing the ledger).

**Review by manish-sethi — COMMENTED** [review]
LGTM except a couple minor comments. 
The test files in kvledger/tests are more readable now, by splitting the tests and moving couch data loading in a separate function.

File: core/ledger/kvledger/upgrade_dbs.go [hunk 8]
```diff
@@ -24,29 +33,44 @@ func UpgradeDBs(rootFSPath string) error {
...
+	if bytes.Equal(format, []byte(dataformat.Version20)) {
+		logger.Info("Ledger data format is current, do nothing.")
+		return nil
+	}
+	if format != nil {
+		err = &dataformat.ErrVersionMismatch{
+			ExpectedVersion: "",
+			Version:         string(format),
+			DBInfo:          fmt.Sprintf("leveldb for channel-IDs at [%s]", dbPath),
+		}
```
[the remark this answers is not in the document]
Comment (cendhu) [reply core/ledger/kvledger/upgrade_dbs.go]: > We had this check deliberately in the function idStore.upgradeFormat. So that we do not accidentally end-up using that function without a precondition check somewhere. I suggest to keep it there for sure. I am fine, if you want to check here again.

I suggested this to be removed from `idStore.upgradeFormat()` because it introduces code duplication, execution duplication, and side-effect in the function `upgradeFormat()`. Currently, `idStore.upgradeFormat()` is used only by the `UpgradeDBs()`. I don't see a strong justification for code duplication, execution duplication, and side-effect but I will step aside if you see it.

[the remark this answers is not in the document]
Comment (manish-sethi) [reply core/ledger/kvledger/upgrade_dbs.go]: I gave the justification in my comment - which is code robustness. A function better not be knowing that how it is being used and hence preconditions be checked inside. 
There are situations when we cannnot avoid that e.g., assume that the lock has been acquired by the caller. But there you still prefer to keep function name either xxxwithoutLock or specify in the comments.
Here, repeating an if check twice in the code flow is not such a big problem if its helping in early return as well as code strength.

What side-effect are you referring to?

[the remark this answers is not in the document]
Comment (cendhu) [reply core/ledger/kvledger/upgrade_dbs.go]: For the above-mentioned justification only, I mentioned it is not strong enough (for me). At the same time, if you are okay with it, I have no problem in having the following.  

- at least 10 lines are duplicated.
- in the UpgradeDBs() execution flow, these 10 lines occurs twice.
- only one functions call updateFormat. Hence, I don't know about the robustness. I also couldn't speculate what other future functions would use it and what are the probabilities that we would miss checking the current format before calling updateFormat. 
- `updateFormat()` is more like `updateFormatIfNotCurrent()`

File: core/ledger/kvledger/upgrade_dbs.go (hunk 6 above)
Comment (wenjianqiao) [reply core/ledger/kvledger/upgrade_dbs.go:42]: I thought this message may alert the user in case the ledger db should not be empty. If this is not the case, we can remove the checking. Otherwise, it is ok for the message to disappear if the user starts the peer once in v2.0 and re-runs upgrade-dbs. We may change the error to a warning message though.

File: core/ledger/kvledger/upgrade_dbs.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (wenjianqiao) [reply core/ledger/kvledger/upgrade_dbs.go]: I moved the checking per review comment mainly because it is a package-only method - we own the caller code to do the checking before calling. However, I agree with Manish to also keep the checking in idStore.upgradeFormat. 
I propose to add a func idStore.checkFormatUpgradeEligibility that will be called by UpgradeDBs and idStore.ugradeFormat. This way idStore.getFormat is not needed any more.

[the remark this answers is not in the document]
Comment (cendhu) [reply core/ledger/kvledger/upgrade_dbs.go]: As the majority feels the other way, I will step aside. Sorry, @wenjianqiao for the extra work incase you have to revert the changes.

File: core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil_test.go [hunk 9]
```diff
@@ -184,3 +186,35 @@ func TestConstructedNamespaceDBName(t *testing.T) {
...
+	couchDBEnv.startCouchDB(t)
+	config.Address = couchDBEnv.couchAddress
+	defer couchDBEnv.cleanup(config)
+	database := "testdropapplicationdbs"
+
+	couchInstance, err := createCouchInstance(config, &disabled.Provider{})
+	assert.NoError(t, err, "Error when trying to create couch instance")
+
+	err = writeDataFormatVersion(couchInstance, dataformat.Version20)
+	assert.NoErrorf(t, err, "Error when trying to write data format version to %s", fabricInternalDBName)
```
[the remark this answers is not in the document]
Comment (wenjianqiao) [reply core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil_test.go]: It was called to create the fabric__internal database. Changed to call createCouchDatabase(couchInstance, fabricInternalDBName) instead.

**Comment by wenjianqiao** [comment]
@manish-sethi @cendhu 
Thank you for your comments. I have updated code to address your comments. The checking for empty db is kept because otherwise we cannot tell why formatKey does not exist (either 1.x version or empty db). Added a new checkUpgradeEligibility func that is called by both UpgradeDBs and upgradeFormat. Removed getFormat.
@manish-sethi  for PreviousVersion and CurrentVersion change, plan to use a separate PR since it impacts other dbs.

File: core/ledger/kvledger/upgrade_dbs.go (hunk 6 above)
Comment (wenjianqiao) [reply core/ledger/kvledger/upgrade_dbs.go]: I keep the check because otherwise we cannot tell why formatKey does not exist (either 1.x version or empty db).

File: core/ledger/kvledger/kv_ledger_provider.go [hunk 10]
```diff
@@ -498,30 +498,51 @@ func openIDStore(path string) (s *idStore, e error) {
...
-		logger.Debug("Format is current, nothing to do")
-		return nil
+	if bytes.Equal(format, []byte(dataformat.Version20)) {
+		logger.Info("Ledger data format is current, nothing to upgrade.")
+		return false, nil
 	}
-	if format != nil {
+	if !bytes.Equal(format, dataformat.Version1xBytes) {
 		err = &dataformat.ErrVersionMismatch{
 			ExpectedVersion: "",
```
Comment (manish-sethi) [inline core/ledger/kvledger/kv_ledger_provider.go:524]: Missed using the constant here but do not want to hold this PR for this.

**Review by manish-sethi — APPROVED** [review]
